model/softfloat/riscv_softfloat_eq.sail

```diff
@@ -0,0 +1,153 @@
+/*=======================================================================================*/
+
+/* **************************************************************** */
+/* This file implements the floating-point for equal and consumes   */
+/* by the softfloat API contract, riscv_softfloat_interface.sail.   */
+/*                                                                  */
+/* The below floating-point types are supported.                    */
+/* 1. Double-precision, aka 64 bits floating-point.                 */
+/* 2. Single-precision, aka 32 bits floating-point.                 */
+/* 3. Half-precision, aka 16 bits floating-point.                   */
+/*                                                                  */
+/* However, the implementation still need some global variables in  */
+/* file riscv_softfloat_interface.sail. It includes:                */
+/* 1. float_result.                                                 */
+/* 2. float_fflags.                                                 */
+/* 3. float_roundingMode.                                           */
+/* **************************************************************** */
+
+val            operator == : (bits_S, bits_S) -> bool
+val            operator |  : (bool, bool) -> bool
+val            operator |  : (bits_S, bits_S) -> bits_S
+val            operator << : (bits_S, int) -> bits_S
+val  f_is_at_least_one_NaN : forall 'm, 'm in {16, 32, 64}. (bits('m), bits('m)) -> bool
+val f_is_at_least_one_SNaN : forall 'm, 'm in {16, 32, 64}. (bits('m), bits('m)) -> bool
+
+val      riscv_softfloat_f16_eq : (bits_H, bits_H) -> unit
+function riscv_softfloat_f16_eq (op1, op2) = {
+  riscv_softfloat_enter (0b000);
+
+  var result : bits_H = zeros(16);
+
+  if f_is_at_least_one_NaN (op1, op2) then {
+    if f_is_at_least_one_SNaN (op1, op2) then {
+      riscv_softfloat_raise_flags (softfloat_flag_invalid);
+    };
+  } else if op1 == op2 | ((op1 | op2) << 1) == zeros(16) then {
+    result[0] = bitone;
+  };
+
+  riscv_softfloat_leave (result);
+}
+
+val      riscv_softfloat_f32_eq : (bits_S, bits_S) -> unit
+function riscv_softfloat_f32_eq (op1, op2) = {
+  riscv_softfloat_enter (0b000);
+
+  var result : bits_S = zeros(32);
+
+  if f_is_at_least_one_NaN (op1, op2) then {
+    if f_is_at_least_one_SNaN (op1, op2) then {
+      riscv_softfloat_raise_flags (softfloat_flag_invalid);
+    };
+  } else if op1 == op2 | ((op1 | op2) << 1) == zeros(32) then {
+    result[0] = bitone;
+  };
+
+  riscv_softfloat_leave (result);
+}
+
+val      riscv_softfloat_f64_eq : (bits_D, bits_D) -> unit
+function riscv_softfloat_f64_eq (op1, op2) = {
+  riscv_softfloat_enter (0b000);
+
+  var result : bits_D = zeros(64);
+
+  if f_is_at_least_one_NaN (op1, op2) then {
+    if f_is_at_least_one_SNaN (op1, op2) then {
+      riscv_softfloat_raise_flags (softfloat_flag_invalid);
+    };
+  } else if op1 == op2 | ((op1 | op2) & 0x7fffffffffffffff) == zeros(64) then {
```

Comment:
Any reason not to use the `<< 1` trick here?

Reply:
Just copy from C side, will try that trick for this.